### ISSUE TITLE
Split up the interval parser

### ISSUE THREAD

**Issue by PeeHaa** [ekinhbayar/IntervalParser#14]
When looking at `IntervalParser::findInterval()` my first instinct is that the stuff in there should be split up into different classes.

Maybe something like:
- Parser
- LeadingParser
- TrailingParser
- BothParser

Cannot think of better names atm. Also not sure if it makes sense what I just said. Will give it some more thought.

**Comment by PeeHaa** [comment]
At the very least it should be split up into separate (private?) methods.

**Comment by ekinhbayar** [comment]
Indeed, it does so many things at the moment. If we are to keep the bitmasked flags those methods shouldn't be allowed for direct use, I think.

**Comment by ekinhbayar** [comment]
Also, I guess another important thing to consider before a real decision here is the hopefully incoming unit tests.

**Comment by PeeHaa** [comment]
Yes. Looks like @memclutter is willing to write them.

Otherwise I will start writing them.

**Comment by ekinhbayar** [comment]
At the moment there is already a `parseInterval()` that returns `DateInterval`. If I take out the `IntervalFlags::INTERVAL_ONLY` bit within `findInterval()` and have it as a solo private method, the only difference between that and the current `parseInterval()` would be the return value. 

The main thought behind having a special `TimeInterval` VO was so that I can hold _extra_ data like leading/trailing, offset and interval length. So considering this, what would you suggest on that case @PeeHaa?

**Comment by ekinhbayar** [comment]
Apart from above, what do you think of renaming `findInterval()` to simply `find()` and splitting the flag based branches of that method into the following private methods:

```
findTimeInterval(string $input): TimeInterval
findIntervalWithLeadingData(string $input): TimeInterval
findIntervalWithTrailingData(string $input): TimeInterval
findAllIntervals(string $input): array
```

Although I'm not sure about the return type of the last one in case it fails to find any, or finds only one.

This is just an idea again. I thought keeping find() as an entry point where passed flags determine expected outcome would be a good option.

**Comment by PeeHaa** [comment]
I just had another look and I think an easy win with fast results is splitting up the `IntervalParser` into three dedicated classes:
- `IntervalFinder`
- `Normalizer`
- `Parser`

Imo these are three separate responsibilities and are easy to refactor out. Once that is done we can have another look into refactoring the interval finder method.

If this also makes sense to you I could refactor it and PR the changes.

**Comment by ekinhbayar** [comment]
Oh this is indeed a great suggestion. It would truly make things easier. Go for it :-)

**Comment by PeeHaa** [comment]
When working on the SRP refactoring I noticed a couple of things:
- The [definition](https://github.com/ekinhbayar/IntervalParser/blob/master/src/IntervalParser.php#L169) is often the same or build up in the same way
- The new `Pattern` class currently just holds some pattern parts, maybe this could be made into a dedicated pattern builder instead.
- There are a couple of places that are currently broken under some circumstances. I.e. there is a possibility that `$intervalOffset` is `null` [here](https://github.com/ekinhbayar/IntervalParser/blob/master/src/IntervalParser.php#L221) because of [this line](https://github.com/ekinhbayar/IntervalParser/blob/master/src/IntervalParser.php#L206) resulting in a type error.
- If we are going to widen the scope of the library we probably want to start sub-namespacing things to keep it organized.

**Comment by ekinhbayar** [comment]
Agreed. Specially on pattern builder, I guess this would also help on a logical separation on ParserSettings, which currently holds the separation expressions and also [modifies](https://github.com/ekinhbayar/IntervalParser/blob/master/src/ParserSettings.php#L66) them. @PeeHaa  do you think having a `build()` method on the new `Pattern` class is a bad idea? Should I have another class for this?
